### PR TITLE
[Mime] Change the way we avoid rendering an email twice

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | yes-ish
| New feature?  | yes-ish 
| Deprecations? | no 
| Tickets       | n/a 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/17065

Trying to kill several birds with one stone here.

I was not very happy with the way we ensured that emails were not renderer more than once.
This solution seems easier and allows for people to render an email **before** sending it to Messenger (important when you have Doctrine entities in the context for instance) as documented in the related doc PR.

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Mime/BodyRenderer.php [hunk 1]
```diff
@@ -64,34 +62,20 @@ public function render(Message $message): void
...
             $message->text($this->twig->render($template, $vars));
+            $message->textTemplate(null);
         }
 
         if ($template = $message->getHtmlTemplate()) {
             $message->html($this->twig->render($template, $vars));
+            $message->htmlTemplate(null);
         }
 
+        $message->context([]);
```
Comment (fabpot) [inline src/Symfony/Bridge/Twig/Mime/BodyRenderer.php:73]: This allows a TemplatedEmail to be serialized after rendering the body when the context contains non-serializable variables like Doctrine entities.

File: src/Symfony/Bridge/Twig/Mime/BodyRenderer.php [hunk 2]
```diff
@@ -45,15 +45,13 @@ public function render(Message $message): void
             return;
         }
 
-        $messageContext = $message->getContext();
-
-        $previousRenderingKey = $messageContext[__CLASS__] ?? null;
-        unset($messageContext[__CLASS__]);
-        $currentRenderingKey = $this->getFingerPrint($message);
-        if ($previousRenderingKey === $currentRenderingKey) {
+        if (null === $message->getTextTemplate() && null === $message->getHtmlTemplate()) {
```
Comment (walva) [inline src/Symfony/Bridge/Twig/Mime/BodyRenderer.php:48]: I would recommend using the method TemplatedEmail::isRendered() the same way it is used [here](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Mailer/Transport/AbstractTransport.php#L83).

Keeping like this prevents subclass of TemplatedEmail with computed html template to not be rendered twice, such as [NotificationEmail](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Bridge/Twig/Mime/NotificationEmail.php#L202).

Comment (OskarStark) [reply src/Symfony/Bridge/Twig/Mime/BodyRenderer.php:48]: Please create a PR or at least a new issue, thanks.

Comment (walva) [reply src/Symfony/Bridge/Twig/Mime/BodyRenderer.php:48]: Thank you for the follow-up @OskarStark, here is the PR : https://github.com/symfony/symfony/pull/59596

**Pull request merged**